### DISCUSSION TITLE
surround: disabling surround type and granular control

### DISCUSSION THREAD

**Discussion by simonmandlik** [nvim-mini/mini.nvim#982] [Q&A] [answered]
Hi,

I would like to

- disable function call `"f"` surround. Tried setting `custom_surroundings = { f = false }` (as in mini.ai), but it throws an error: `Error executing lua vim/shared.lua:0: after the second argument: expected table, got boolean`
- remap original function call to `"C"`, but use the treesitter gen_spec for all operations except `sa`, where it fails. In such case I would like to use function call as provided by surround.mini.

Is such fine-grained control possible? Thanks!

**Comment by simonmandlik** [comment]
Found the answer for the first point. Not sure if it is recommended, but this will work:

```lua
custom_surroundings = {
  f = { input = {}, output = {} },
}
```

**Comment by echasnovski** [comment, marked as answer]
> * disable function call `"f"` surround. Tried setting `custom_surroundings = { f = false }` (as in mini.ai), but it throws an error: `Error executing lua vim/shared.lua:0: after the second argument: expected table, got boolean`

Disabling built-in surrounding identifiers doesn't make much sense, as there is typically nothing to fall back to. It makes sense in 'mini.ai', because there are many built-in `a` / `i` textobjects (like `i)`, `a"`, etc.), not even mentioning coming from other plugins. So the best suggestion for "disabling" is indeed overriding it with something else. Empty tables as specifications is not the case that is explicitly supported, though. So maybe coming up with something else is a good idea.

> * remap original function call to `"C"`, but use the treesitter gen_spec for all operations except `sa`, where it fails. In such case I would like to use function call as provided by surround.mini.

Surrounding specification has exactly this kind of separation into [`input` and `output` surroundings](https://github.com/echasnovski/mini.nvim/blob/ebc8cb0c1f73ee6c53e0560957a37a2307a70f98/doc/mini-surround.txt#L262-L265). This use case can be achieved by using [`gen_spec.input.treesitter()`](https://github.com/echasnovski/mini.nvim/blob/a3f274633a57c713160baf51724ecfc051117877/doc/mini-surround.txt#L775-L783) as input part (the linked example was previously a bit earlier in the help file while the example in `gen_spec.input.treesitter()` help was flawed; fixed it), while output can use `user_input()` (exported specifically for this kind of overrides). Here is the full setup:

```lua
local ts_input = require('mini.surround').gen_spec.input.treesitter
require('mini.surround').setup({
  custom_surroundings = {
    C = {
      input = ts_input({ outer = '@call.outer', inner = '@call.inner' }),
      output = function()
        local fun_name = MiniSurround.user_input('Function name')
        if fun_name == nil then return nil end
        return { left = fun_name .. '(', right = ')' }
      end,
    },
  },
})
```

**Reply by simonmandlik** [reply]
@echasnovski Thanks for the help!

I have noticed that in this config `srC` doesn't work. But it looks like that it doesn't even in the default setting. Is this intended behavior?

**Reply by echasnovski** [reply]
`srC` does work as expected with this config. For example, `srC)` finds function call based on `@call.outer` and `@call.inner` tree-sitter queries and replaces it with parenthesis.

Do you have necessary queries set up [as described in help page](https://github.com/echasnovski/mini.nvim/blob/main/doc/mini-surround.txt#L752-L770)? Usually it means just installing 'nvim-treesitter/nvim-treesitter-textobjects' (without needing to "activate" it).

**Reply by simonmandlik** [reply]
Apologies, I wasn't specific enough. `srC` is working for me, but allows me to only replace the function call with a single-character surround, not rename the function call.

**Reply by echasnovski** [reply]
"Replace" operator requires two identifiers: first to find the one to be replaced, second to define with which it should be replaced (as [shown in examples](https://github.com/echasnovski/mini.nvim/blob/a3f274633a57c713160baf51724ecfc051117877/doc/mini-surround.txt#L77)). So to rename a function call, it is `srCC` - replace function call with another function call (which will show prompt for function name). This is by design to allow more flexible "replace" operator (instead of "rename").

**Reply by simonmandlik** [reply]
thanks!
